From 13b0e99a507eb399ed722dbc323b2f801c9879dd Mon Sep 17 00:00:00 2001
From: Anish Trivedi <anish@freescale.com>
Date: Fri, 7 May 2010 10:36:38 -0500
Subject: [PATCH 154/442] ENGR00123281 SRTC: Add test application to test time
 change notification

Created test to use RTC_WAIT_FOR_TIME_SET ioctl

Signed-off-by: Anish Trivedi <anish@freescale.com>
---
 test/mxc_rtc/Makefile                       |   2 +-
 test/mxc_rtc/autorun-rtc.sh                 |   2 +
 test/mxc_rtc/rtc_timesetnotification_test.c | 235 ++++++++++++++++++++++++++++
 3 files changed, 238 insertions(+), 1 deletion(-)
 create mode 100644 test/mxc_rtc/rtc_timesetnotification_test.c

diff --git a/test/mxc_rtc/Makefile b/test/mxc_rtc/Makefile
index e3bcdb5..12eea11 100644
--- a/test/mxc_rtc/Makefile
+++ b/test/mxc_rtc/Makefile
@@ -2,7 +2,7 @@
 EXCLUDE_LIST:= IMX25_3STACK
 
 ifeq (,$(findstring $(PLATFORM), $(EXCLUDE_LIST)))
-OBJS = $(OBJDIR)/rtctest.out $(OBJDIR)/rtcwakeup.out $(OBJDIR)/autorun-rtc.sh
+OBJS = $(OBJDIR)/rtc_timesetnotification_test.out $(OBJDIR)/rtctest.out $(OBJDIR)/rtcwakeup.out $(OBJDIR)/autorun-rtc.sh
 else
 OBJS =
 endif
diff --git a/test/mxc_rtc/autorun-rtc.sh b/test/mxc_rtc/autorun-rtc.sh
index e9789a6..d15b945 100755
--- a/test/mxc_rtc/autorun-rtc.sh
+++ b/test/mxc_rtc/autorun-rtc.sh
@@ -39,6 +39,8 @@ if [ "$RTC_IRQS" != "$RTC_IRQS_EXPECTED" ]; then
 	STATUS=1
 else
 	echo "checking rtc interrupts PASS"
+       #RTC wait for time set notification test
+        run_testcase "./rtc_timesetnotification_test.out"
 fi
 
 print_status
diff --git a/test/mxc_rtc/rtc_timesetnotification_test.c b/test/mxc_rtc/rtc_timesetnotification_test.c
new file mode 100644
index 0000000..4ca2d97
--- /dev/null
+++ b/test/mxc_rtc/rtc_timesetnotification_test.c
@@ -0,0 +1,235 @@
+/*
+ * Copyright (C) 2010 Freescale Semiconductor, Inc. All rights reserved.
+ */
+
+/*
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+/*
+ *	Real Time Clock Driver Test/Example Program
+ *
+ * rtc_wait_for_time_set_test -- test & demo the use of
+ * RTC_WAIT_FOR_TIME_SET and RTC_READ_TIME_47BIT
+ * ioctls in SRTC driver.
+ *
+ * The child process calls RTC_WAIT_FOR_TIME_SET ioctl,
+ * which is a blocking call until the time is set by the parent process.
+ * The blocking ioctl returns upon time set with a difference
+ * between new time counter value and old time counter value.
+ * Processes such as DRM can utilize these ioctls to keep track of
+ * changes to the SRTC in order to implement a secure clock:
+ * secure clock value = SRTC counter value + offset, where offset
+ * is kept in sync with any changes made to the SRTC clock counter.
+ *
+ * The parent process, in this test, first sets time ahead by 16 seconds
+ * and then sets the time back by 16 seconds. Each time, the child process
+ * is woken up, and it returns failure if the time difference is not +16 seconds
+ * and -16 seconds, respectively. When time is changed by seconds, we check
+ * for rollover in the minute and hour fields. However, rollover into the day field
+ * is not checked. Therefore, if this test is run close to midnight, then it may fail.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <time.h>
+
+#include <sys/ioctl.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include <linux/rtc.h>
+#include <linux/mxc_srtc.h>
+
+int main(int argc, char **argv)
+{
+	int fd_srtc;
+	unsigned long long time_47bit;
+	long long time_diff;
+	struct rtc_time rtc_tm;
+	int retval = 0;
+
+	fd_srtc = open("/dev/rtc0", O_RDONLY);
+
+	if (fd_srtc == -1) {
+		perror("/dev/rtc0");
+		exit(errno);
+	}
+
+	fprintf(stderr, "\n\t\t\tRTC_WAIT_FOR_TIME_SET Test.\n\n");
+
+
+	/* first, let's just read the time from SRTC */
+	retval = ioctl(fd_srtc, RTC_RD_TIME, &rtc_tm);
+	if (retval == -1) {
+		perror("Error! RTC_RD_TIME");
+		exit(errno);
+	}
+
+	/* Now, let's read the whole 47-bit counter value */
+	retval = ioctl(fd_srtc, RTC_READ_TIME_47BIT, &time_47bit);
+	if (retval == -1) {
+		perror("Error! RTC_READ_TIME_47BIT");
+		exit(errno);
+	}
+
+	fprintf(stderr, "Current RTC date/time is %d-%d-%d, %02d:%02d:%02d.\n",
+		rtc_tm.tm_mday, rtc_tm.tm_mon + 1, rtc_tm.tm_year + 1900,
+		rtc_tm.tm_hour, rtc_tm.tm_min, rtc_tm.tm_sec);
+
+
+	fprintf(stderr, "47-bit Time counter on SRTC = 0x%llx\n", time_47bit);
+
+	/* fork a child process that will call the ioctl that waits for time set */
+	pid_t pID = fork();
+
+	/* parent process */
+	if (pID > 0) {
+
+		/* Set the time to 16 sec ahead, and check for rollover of min & hr.
+		 * Does not check for rollover of days, so this test may fail
+		 * when run with RTC time = 11:59 PM */
+		rtc_tm.tm_sec += 16;
+		if (rtc_tm.tm_sec >= 60) {
+			rtc_tm.tm_sec %= 60;
+			rtc_tm.tm_min++;
+		}
+		if (rtc_tm.tm_min == 60) {
+			rtc_tm.tm_min = 0;
+			rtc_tm.tm_hour++;
+		}
+		if (rtc_tm.tm_hour == 24)
+			rtc_tm.tm_hour = 0;
+
+
+		fprintf(stderr, "Trying to set RTC date/time to %d-%d-%d, %02d:%02d:%02d.\n",
+			rtc_tm.tm_mday, rtc_tm.tm_mon + 1, rtc_tm.tm_year + 1900,
+			rtc_tm.tm_hour, rtc_tm.tm_min, rtc_tm.tm_sec);
+
+		retval = ioctl(fd_srtc, RTC_SET_TIME, &rtc_tm);
+		if (retval == -1) {
+			perror("Error! RTC_SET_TIME");
+			exit(errno);
+		}
+
+		/* check to see if time was indeed set ahead by close to 16 seconds */
+		retval = ioctl(fd_srtc, RTC_RD_TIME, &rtc_tm);
+		if (retval == -1) {
+			perror("Error! RTC_RD_TIME");
+			exit(errno);
+		}
+
+		fprintf(stderr, "Current RTC date/time is %d-%d-%d, %02d:%02d:%02d.\n",
+			rtc_tm.tm_mday, rtc_tm.tm_mon + 1, rtc_tm.tm_year + 1900,
+			rtc_tm.tm_hour, rtc_tm.tm_min, rtc_tm.tm_sec);
+
+
+		/* Set the time to 16 sec back, and check for rollover of min & hr.
+		 * Does not check for rollover of days, so this test may fail
+		 * when run with RTC time = 12:00 AM */
+		if (rtc_tm.tm_sec > 15)
+			rtc_tm.tm_sec -= 16;
+		else
+			rtc_tm.tm_sec = 44 + rtc_tm.tm_sec;
+		if (rtc_tm.tm_sec >= 44) {
+			if (rtc_tm.tm_min != 0)
+				rtc_tm.tm_min--;
+			else
+				rtc_tm.tm_min = 59;
+		}
+		if (rtc_tm.tm_min == 59 && rtc_tm.tm_sec >= 44) {
+			if (rtc_tm.tm_hour != 0)
+				rtc_tm.tm_hour--;
+			else
+				rtc_tm.tm_hour = 23;
+		}
+
+		fprintf(stderr, "Trying to set RTC date/time to %d-%d-%d, %02d:%02d:%02d.\n",
+			rtc_tm.tm_mday, rtc_tm.tm_mon + 1, rtc_tm.tm_year + 1900,
+			rtc_tm.tm_hour, rtc_tm.tm_min, rtc_tm.tm_sec);
+
+		retval = ioctl(fd_srtc, RTC_SET_TIME, &rtc_tm);
+		if (retval == -1) {
+			perror("Error! RTC_SET_TIME");
+			exit(errno);
+		}
+
+		/* check to see if time was indeed set back by close to 16 seconds */
+		retval = ioctl(fd_srtc, RTC_RD_TIME, &rtc_tm);
+		if (retval == -1) {
+			perror("Error! RTC_RD_TIME");
+			exit(errno);
+		}
+
+		fprintf(stderr, "Current RTC date/time is %d-%d-%d, %02d:%02d:%02d.\n",
+			rtc_tm.tm_mday, rtc_tm.tm_mon + 1, rtc_tm.tm_year + 1900,
+			rtc_tm.tm_hour, rtc_tm.tm_min, rtc_tm.tm_sec);
+
+
+		/* wait for child to complete before exiting, and get its return value */
+		wait(&retval);
+		retval = WEXITSTATUS(retval);
+
+	}
+	/* child process */
+	else if (pID == 0) {
+
+		/* this is a blocking call */
+		retval = ioctl(fd_srtc, RTC_WAIT_TIME_SET, &time_diff);
+		if (retval == -1) {
+			perror("Child process: Error! RTC_WAIT_TIME_SET");
+			exit(errno);
+		}
+
+		/* signed value means add 1 second */
+		time_diff += (1 << 15);
+
+		fprintf(stderr, "Child process: time_diff = 0x%llx = %d seconds\n", time_diff, (int) (time_diff >> 15));
+
+		/* The difference should be +16 seconds */
+		if ((time_diff >> 15) != 16) {
+			fprintf(stderr, "Child process: Error! time_diff was not reported as 16 seconds\n");
+			exit(1);
+		}
+
+		/* this is a blocking call */
+		retval = ioctl(fd_srtc, RTC_WAIT_TIME_SET, &time_diff);
+		if (retval == -1) {
+			perror("Child process: Error! RTC_WAIT_TIME_SET");
+			exit(errno);
+		}
+
+		/* signed value means add 1 second */
+		time_diff += (1 << 15);
+
+		fprintf(stderr, "Child process: time_diff = 0x%llx = %d seconds\n", time_diff, 	(int) (time_diff >> 15));
+
+		/* The difference should be -16 seconds */
+		if ((time_diff >> 15) != -16) {
+			fprintf(stderr, "Child process: Error! time_diff was not reported as -16 seconds\n");
+			exit(1);
+		}
+
+		exit(0);
+	}
+	/* error during fork */
+	else {
+		perror("Error: Could not fork a child");
+		close(fd_srtc);
+		exit(errno);
+	}
+
+	close(fd_srtc);
+	return retval;
+}
+
-- 
1.8.0

